tools/cldr-code/src/main/java/org/unicode/cldr/tool/LangTagsData.java

```diff
@@ -0,0 +1,387 @@
+                        CLDRLocale targetForLanguage = fullForLanguage.getCldrLocale();
+                        CLDRLocale target = entry.getValue().getCldrLocale();
+                        // The missing value in LSRSource must not be the same as what would come in
+                        // that is, if we have aaa => aaa_Bbbb_CC, then we cannot have:
+                        // aaa_Dddd => aaa_Dddd_CC, nor
+                        // aaa_EE => aaa_Bbbb_EE, nor
+                        if (target.getLanguage().equals(targetForLanguage.getLanguage())
+                                || target.getScript().equals(targetForLanguage.getScript())) {
+                            toRemove.add(source);
+                            processErrors.put(
+                                    Errors.Type.redundant_mapping,
+                                    source,
+                                    entry.getValue().getLsrString(),
+                                    "because: " + language + " ➡ " + targetForLanguage);
+                        }
+                    }
+                }
+            }
+            for (String badKey : toRemove) {
+                result.remove(badKey);
+            }
+
+            // protect the results
+
+            processErrors.data = CldrUtility.protectCollection(processErrors.data);
+            return CldrUtility.protectCollection(result);
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    private boolean isIllFormed(String source, LanguageTagParser languageTagParser) {
+        return languageTagParser.getLanguage().isEmpty()
+                || !languageTagParser.getVariants().isEmpty()
+                || !languageTagParser.getExtensions().isEmpty()
+                || !languageTagParser.getLocaleExtensions().isEmpty()
+                || source.contains("@");
+    }
+
+    private boolean isOk(String lang, String script, String region, Map<LstrType, Status> errors) {
+        errors.clear();
+        if (!LIKELY_SPECIALS.contains(lang)) {
+            check(LstrType.language, lang, errors);
+        }
+        if (!FIX_VALIDITY.contains(script)) {
+            check(LstrType.script, script, errors);
+        }
+        if (region.equals("001") && Iso639Data.getType(lang) == Type.Constructed) {
+            // ok
+        } else {
+            check(LstrType.region, region, errors);
+        }
+        return errors.isEmpty();
+    }
+
+    private void check(LstrType lstrType, String lang, Map<LstrType, Status> errors) {
+        final Status status = validity.getCodeToStatus(lstrType).get(lang);
+        if (status != Status.regular) {
+            errors.put(lstrType, status);
+        }
+    }
+
+    private Multimap<String, String> readWikidata() {
+        Multimap<String, String> result = TreeMultimap.create();
+        Path path =
+                Paths.get(CLDRPaths.BIRTH_DATA_DIR, "/../external/wididata_lang_region.tsv")
```

Comment:
perhaps rename the file too at some point?

```suggestion
                Paths.get(CLDRPaths.BIRTH_DATA_DIR, "/../external/wikidata_lang_region.tsv")
```

Reply:
good idea. Actually, we should be using a tool to get that data from wikidata; right now it was just a one-time query.